src/connectors/tests/gpubsub/gpub.rs

```diff
@@ -0,0 +1,296 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::connectors::impls::gpubsub::producer::Builder;
+use crate::connectors::tests::ConnectorHarness;
+use crate::connectors::utils::EnvHelper;
+use crate::errors::Result;
+use crate::instance::State;
+use async_std::prelude::FutureExt;
+use googapis::google::pubsub::v1::publisher_client::PublisherClient;
+use googapis::google::pubsub::v1::subscriber_client::SubscriberClient;
+use googapis::google::pubsub::v1::{PullRequest, Subscription, Topic};
+use serial_test::serial;
+use std::collections::HashSet;
+use std::time::Duration;
+use testcontainers::clients::Cli;
+use testcontainers::RunnableImage;
+use tonic::transport::Channel;
+use tremor_common::ports::IN;
+use tremor_pipeline::{Event, EventId};
+use tremor_value::{literal, Value};
+
+#[async_std::test]
+#[serial(gpubsub)]
+async fn no_connection() -> Result<()> {
+    let _ = env_logger::try_init();
+    let connector_yaml = literal!({
+        "codec": "binary",
+        "config":{
+            "endpoint": "https://localhost:9090",
+            "connect_timeout": 100000000,
+            "topic": "projects/xxx/topics/test-a",
+            "skip_authentication": false
+        }
+    });
+
+    let harness =
+        ConnectorHarness::new(function_name!(), &Builder::default(), &connector_yaml).await?;
+    assert!(harness.start().await.is_err());
+    Ok(())
+}
+
+#[async_std::test]
+#[serial(gpubsub)]
+async fn no_hostname() -> Result<()> {
+    let _ = env_logger::try_init();
+    let connector_yaml = literal!({
+        "codec": "binary",
+        "config":{
+            "endpoint": "file:///etc/passwd",
+            "connect_timeout": 100000000,
+            "topic": "projects/xxx/topics/test-a",
+            "skip_authentication": false
+        }
+    });
+
+    let harness =
+        ConnectorHarness::new(function_name!(), &Builder::default(), &connector_yaml).await;
+
+    assert!(harness.is_err());
+
+    Ok(())
+}
+
+#[async_std::test]
+#[serial(gpubsub)]
+async fn no_token() -> Result<()> {
+    let _ = env_logger::try_init();
+
+    let runner = Cli::docker();
+
+    let (pubsub, pubsub_args) =
+        testcontainers::images::google_cloud_sdk_emulators::CloudSdk::pubsub();
+    let runnable_image = RunnableImage::from((pubsub, pubsub_args));
+    let container = runner.run(runnable_image);
+
+    let port = container
+        .get_host_port_ipv4(testcontainers::images::google_cloud_sdk_emulators::PUBSUB_PORT);
+    let endpoint = format!("http://localhost:{}", port);
+
+    let mut env = EnvHelper::new();
+    env.remove_var("GOOGLE_APPLICATION_CREDENTIALS");
```

Comment:
This test might fail if we run this test on a gcloud VM or there is a credentials file somewhere where `gcloud` expects it. According to the documentation here: https://crates.io/crates/gouth Is this correct?

Reply:
true,  though I'm not sure if there's a foolproof way to enforce a failure to find the token, so maybe it's a risk worth taking

Reply:
You could, at the start of the test try to get a token manually and only run the test and the asserts in it if this was failing, otherwise "skip" it with just an `Ok(())`.

Reply:
But I'm afraid that if we do that, then if the tests end up running on GCP, it will not work and just go unnoticed. It will fails, it will be mildly annoying, but at least we will know that action needs to be taken.

Reply:
Ok, lets add a hook that is calling you whenever this test fails. :D